File: java/374.猜数字大小.java
/*
 * @lc app=leetcode.cn id=374 lang=java
 *
 * [374] 猜数字大小
 */

// @lc code=start
/**
 * Forward declaration of guess API.
 * 
 * @param num your guess
 * @return -1 if num is lower than the guess number
 *         1 if num is higher than the guess number
 *         otherwise return 0
 *         int guess(int num);
 */

public class Solution extends GuessGame {
    public int guessNumber(int n) {
        int l = 1, r = n;
        while (l < r) {
            int m = (r - l) / 2 + l;
            int result = guess(m);
            if (result == -1) {
                r = m - 1;
            } else if (result == 1) {
                l = m + 1;
            } else {
                return m;
            }
        }
        return l;
    }
}
// @lc code=end

File: solution/python/374.猜数字大小.py
#
# @lc app=leetcode.cn id=374 lang=python3
# @lcpr version=30203
#
# [374] 猜数字大小
#

import sys
import os

sys.path.append(os.path.join(os.path.dirname(__file__), ".."))

from typing import *
from common.node import *

SECRET = 1


def guess(num: int) -> int:
    if num == SECRET:
        return 0
    return -1 if num > SECRET else 1


# @lc code=start
class Solution:
    def guessNumber(self, n: int) -> int:
        left, right = 1, n
        while left <= right:
            mid = (left + right) // 2
            res = guess(mid)
            if res == 0:
                return mid
            if res < 0:
                right = mid - 1
            else:
                left = mid + 1
        return -1
        # @lc code=end


if __name__ == "__main__":
    solution = Solution()
    # 测试用例 (func, args, result)
    test_cases = [
        (solution.guessNumber, [10], 6),
        (solution.guessNumber, [1], 1),
        (solution.guessNumber, [2], 2),
    ]

    all_passed = True
    for idx, (func, args, expected) in enumerate(test_cases):
        try:
            SECRET = expected
            result = func(*args)
            assert result == expected
            print(f"测试用例 {idx + 1} 通过: n = {args}, result = {result}")
        except AssertionError:
            all_passed = False
            print(
                f"测试用例 {idx + 1} 失败: n = {args}, 期望 = {expected}, 实际 = {result}"
            )

    file_path = os.path.basename(__file__).split(".")
    file_number = file_path[0]
    file_name = file_path[1]
    if all_passed:
        print(f'第 {file_number} 题 "{file_name}" 所有测试用例通过')
        sys.exit(0)
    else:
        print(f'第 {file_number} 题 "{file_name}" 部分测试用例失败')
        sys.exit(1)


#
# @lcpr case=start
# 10\n
# @lcpr case=end

#
